[CS-4060] Hide Welcome screen when the app restarts

paulinhapenedo opened the pull request:
### Description
~~This PR doesn't solve the issue described in the ticket, but I think it helps with the "awkwardness" of restarting the app while being authenticated and seeing briefly the Welcome screen.~~

This PR replaces the library we've been using to handle the Splash Screen to one that's being actively maintained (`react-native-bootsplash`) in a fixed version (3.6.2) so we can show and hide the splash screen programatically. A new version of this library (4.0) removes the `show` API because there's no native support for this on Android 12.

- [x] Completes #CS-4060

### Checklist

- [x] Tested on a small device
- [x] Tested on iOS
- [x] Tested on Android

### Screenshots
https://user-images.githubusercontent.com/690904/176951954-309ae10a-e492-450e-b7b7-988dfb3d1a3b.mov

https://user-images.githubusercontent.com/690904/176951968-181649eb-e0dc-465a-870c-9673c95239a6.mp4

paulinhapenedo committed:
[CS-4060] Hide Welcome screen when the app restarts

oidoug commented:
Maybe the problem is that we are hiding the splash too soon, before even a wallet has been detected so we go straight to unauthenticated flow.

Your fix makes it better than before but I'm not sure it's the right answer, so I'll let Dani give her opinion on before approving.

danibonilha reviewed (commented):
I believe a more user friendly approach would be to show the splash screen after calling `Restart`, I'm aware that `SplashScreen.show()` doesn't seem to work on iOS, but I think it's worth investigating why, we can timebox to one day or so, I can help out if you need, lmk.

paulinhapenedo committed:
fix: replace react-native-splash-screen for react-native-bootsplash lib

paulinhapenedo commented:
# Update

Replaced `react-native-splash-screen` that's not being actively maintained with `react-native-bootsplash` that has a good support. We've fixed version on 3.6.2 because we need the `show` method. After the version 4.0, the maintainer removed that because there's no native support on Android.

The first splash screen on Android, the one with a white background, will be handled in another PR to broadening the scope of this one.

## Android cold start
https://user-images.githubusercontent.com/690904/177347857-a80d7094-4e28-463a-8c74-6f5b07938fdc.mov

## Android restart after something went wrong
https://user-images.githubusercontent.com/690904/177347887-54f1e66b-636d-4891-a623-39182fe425f2.mov

## iOS cold start
https://user-images.githubusercontent.com/690904/177350603-c1e0906a-9162-4062-9604-bf9706f7d5ab.mp4

## iOS restart after something went wrong
https://user-images.githubusercontent.com/690904/177350621-c2b9b1d9-bded-4012-a011-4d7af8c6142b.mp4

danibonilha commented on cardstack/src/navigation/tabBarNavigator.tsx:
```diff
@@ -191,11 +191,14 @@ const stackScreenOptions: StackNavigationOptions = {
   return (
     <Stack.Navigator screenOptions={stackScreenOptions}>
-      {!hasWallet ? (
+      {showWelcomeScreen ? (
```
Do we still need this ?

paulinhapenedo replied:
No, removing it.

danibonilha commented on src/components/settings-menu/DeveloperSettings.js:
```diff
@@ -53,7 +54,13 @@ const DeveloperSettings = () => {
+        label="🔄 Restart app"
+        onPress={() => {
+          RNRestart.Restart();
+          RNBootSplash.show();
```
To avoid forgetting that right after a Restart we also need to call the splashScreen we could create a helper function to do that, this way we can use only one function across the app

danibonilha commented on android/app/src/main/res/drawable/bootsplash.xml:
```diff
@@ -0,0 +1,9 @@
+    <item android:drawable="@color/black" />
+
+    <item>
+        <bitmap android:src="@mipmap/bootsplash_logo" android:gravity="center" />
```
we have the app_logo_splash, which is pretty much the same as the bootsplash_logo, so we should remove the app_logo_splash, alongside with the file that's using it, `launch_screen.xml` since it's been replaced with this one

danibonilha commented on ios/Rainbow/Images.xcassets/BootSplashLogo.imageset/Contents.json:
```diff
@@ -0,0 +1,23 @@
+    "version": 1,
+    "author": "xcode"
+  }
+}
```
We already have the logo, it's `CardstackIcon` so we probably don't need to add this new one

danibonilha commented on ios/Rainbow/BootSplash.storyboard:
```diff
@@ -0,0 +1,45 @@
+    <resources>
+        <image name="BootSplashLogo" width="100" height="92"/>
+    </resources>
+</document>
```
I'm confused on why we have a `LauchScreen.storyboard` and also this file that seems pretty much the same thing. it seems this BootSplash.storyboard, is not even being used, given that `LaunchScreen.storyboard ` is the one added on project.pbxproj.

paulinhapenedo replied:
This was created by the command line interface the new library provides to create the assets and the splash screen. Will remove this.

paulinhapenedo committed:
fix: revert unnecessary files

paulinhapenedo committed:
feat: create utils function to restart the app showing the splash screen

oidoug reviewed (commented):
Code looks good, but I can't see the splash on my device, just a blank screen for a couple of secs.

https://user-images.githubusercontent.com/129619/177549055-93982eb6-4813-42e1-b9c5-06c775e4664d.mp4

oidoug reviewed (approved):
I re-re-visited this, and it is working as expected. I didn't have the latest commits 😓, sorry!

One minor thing to consider if it's possible to fix or not. On iOS the transition is pretty smooth, but on android it seems the Status Bar height changes and it makes the splash to jump a bit, as shown in one of the screenshots - not a blocker though.

paulinhapenedo committed:
fix: change Android 12 Splash screen background color

danibonilha reviewed (approved):
LGTM, just make sure to test android's deep linking before merging it 😉
